pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,355 @@
+package dockerfile
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/build"
+	s2ierr "github.com/openshift/source-to-image/pkg/errors"
+	"github.com/openshift/source-to-image/pkg/ignore"
+	"github.com/openshift/source-to-image/pkg/scm"
+	"github.com/openshift/source-to-image/pkg/scm/git"
+	"github.com/openshift/source-to-image/pkg/scripts"
+	"github.com/openshift/source-to-image/pkg/util"
+	"github.com/openshift/source-to-image/pkg/util/fs"
+	utilglog "github.com/openshift/source-to-image/pkg/util/glog"
+	utilstatus "github.com/openshift/source-to-image/pkg/util/status"
+	"github.com/openshift/source-to-image/pkg/util/user"
+)
+
+const defaultDestination = "/tmp"
+
+var (
+	glog = utilglog.StderrLog
+
+	// List of directories that needs to be present inside working dir
+	workingDirs = []string{
+		api.UploadScripts,
+		api.Source,
+		api.DefaultScripts,
+		api.UserScripts,
+	}
+)
+
+// Dockerfile builders produce a Dockerfile rather than an image.
+// Building the dockerfile w/ the right context will result in
+// an application image being produced.
+type Dockerfile struct {
+	fs               fs.FileSystem
+	uploadScriptsDir string
+	uploadSrcDir     string
+	sourceInfo       *git.SourceInfo
+	result           *api.Result
+	ignorer          build.Ignorer
+}
+
+// New creates a Dockerfile builder.
+func New(config *api.Config, fs fs.FileSystem) (*Dockerfile, error) {
+	return &Dockerfile{
+		fs: fs,
+		// where we will get the assemble/run scripts from on the host machine,
+		// if any are provided.
+		uploadScriptsDir: api.UploadScripts,
+		uploadSrcDir:     api.Source,
+		result:           &api.Result{},
+		ignorer:          &ignore.DockerIgnorer{},
+	}, nil
+}
+
+// Build produces a Dockerfile that when run with the correct filesystem
+// context, will produce the application image.
+func (builder *Dockerfile) Build(config *api.Config) (*api.Result, error) {
+
+	// Handle defaulting of the configuration that is unique to the dockerfile strategy
+	if strings.HasSuffix(config.AsDockerfile, string(os.PathSeparator)) {
+		config.AsDockerfile = config.AsDockerfile + string(os.PathSeparator) + "Dockerfile"
+	}
+	if len(config.AssembleUser) == 0 {
+		config.AssembleUser = "1001"
+	}
+	if !user.IsUserAllowed(config.AssembleUser, &config.AllowedUIDs) {
+		return nil, s2ierr.NewUserNotAllowedError(config.AssembleUser, false)
```

Comment:
i doubt we're going to re-check the dockerfile downstream.  that said:

1) the label logic would be irrelevant here since we have no way to inspect the image to look for one..the assumption is the upstream logic would detect such a label and explicitly pass the assemble-user arg in the config

2) if the base image has ONBUILD commands, those would also need to be validated upstream because we have no way to inspect the image here.  But they do need to be validated because "dockerfile output" is actually a lot like a layered build.